File: cpp/3.cpp
#include <iostream>
#include <unordered_set>

using namespace std;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    int ret = 0;
    int l = -1, r = 0;
    unordered_set<char> seen;
    while (r < s.size()) {
      while (seen.find(s[r]) != seen.end())
        seen.erase(s[++l]);
      r - l > ret ? ret = r - l : ret = ret;
      seen.insert(s[r++]);
    }
    return ret;
  }
};

File: python/3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ret, l, r = 0, -1, 1
        seen = set()
        while r < len(s):
            while s[r] in seen:
                l += 1
                seen.remove(s)
        return ret
